### PR TITLE
Form helper

### PR DESCRIPTION
**Pull request by Portaflex**
Form helper without Form Validation. Test passed.

### CODE REVIEW

File: system/Helpers/form_helper.php [hunk 1]
```diff
@@ -0,0 +1,859 @@
...
+        $form = '<form action="'.$action.'"'.$attributes.">\n";
+        
+        // Add CSRF field if enabled, but leave it out for GET requests and requests to external websites
+        $getCsrf = new \Config\Filters();
+        $before = $getCsrf->globals->before;
+        if ((in_array('csrf', $before) || array_key_exists('csrf', $before)) &&
+                strpos($action, base_url()) !== false && ! stripos($form, 'method="get"'))
+        {
+            $security = \Config\Services::security();
+            $hidden[$security->getCsrfTokenName()] = $security->getCsrfHash();
```
Comment (lonnieezell) [inline system/Helpers/form_helper.php]: Methods names need to have CSRF capitalized: `getCSRFTokenName()` and `getCSRFHash()`.

File: system/Helpers/form_helper.php [hunk 2]
```diff
@@ -0,0 +1,859 @@
...
+     * @param	string
+     * @param	mixed
+     * @return	string
+     */
+    function form_upload($data = '', string $value = '', $extra = ''): string
+    {
+        $defaults = ['type' => 'file', 'name' => ''];
+        is_array($data) OR $data = ['name' => $data];
+        $data['type'] = 'file';
+        return '<input '._parse_form_attributes($data, $defaults)._attributes_to_string($extra)." />\n";
```
Comment (lonnieezell) [inline system/Helpers/form_helper.php]: Please include blank lines between logical groups within code for better scanability/readability.

File: system/Helpers/form_helper.php [hunk 3]
```diff
@@ -0,0 +1,859 @@
...
+    /**
+     * Drop-down Menu
+     *
+     * @param	mixed	$data
+     * @param	mixed	$options
+     * @param	mixed	$selected
+     * @param	mixed	$extra
+     * @return	string
+     */
+    function form_dropdown($data = '', $options = [], $selected = [], $extra = ''): string
```
Comment (lonnieezell) [inline system/Helpers/form_helper.php]: You forgot type hints on this method.

File: system/Helpers/form_helper.php [hunk 4]
```diff
@@ -0,0 +1,859 @@
...
+    /**
+     * Checkbox Field
+     *
+     * @param	mixed
+     * @param	string
+     * @param	bool
+     * @param	mixed
+     * @return	string
+     */
+    function form_checkbox($data = '', string $value = '', bool $checked = false, $extra = ''): string
```
Comment (lonnieezell) [inline system/Helpers/form_helper.php]: Missing a couple of type hints. Please verify that all methods use type hints where possible.

File: system/Helpers/form_helper.php [hunk 5]
```diff
@@ -0,0 +1,859 @@
...
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_keygen'))
+{
+    function form_keygen()
+    {
+        /* The purpose of the <keygen> element is to provide a secure way to authenticate users.
```
Comment (lonnieezell) [inline system/Helpers/form_helper.php]: Missing function body.

File: system/Helpers/form_helper.php [hunk 6]
```diff
@@ -0,0 +1,859 @@
...
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_output'))
+{
+    function form_output()
+    {
+        /* The <output> element represents the result of a calculation (like one performed by a script). */
```
Comment (lonnieezell) [inline system/Helpers/form_helper.php]: Missing function body

**Comment by lonnieezell** [comment]
I see you've started to include some of the newer HTML5 inputs, which is great (though at least a couple are unimplmented at the moment). However, I didn't see some of the most useful new ones like email, number, date, datetime, and phone number fields.

File: system/Helpers/form_helper.php (hunk 3 above)
Comment (Portaflex) [inline system/Helpers/form_helper.php]: Didn't type hint this one because the inputs are mixed, strings or arrays.

**Comment by Portaflex** [comment]
Ok. If we talk about new HTML Form Elements, those are datalist, keygen and output, no more. If we talk about Input types then we have all those that you talk about. So, as we have it now, we need just to pass the 'type' data and some other extras to form_input(). Thanks.

**Comment by lonnieezell** [comment]
@Portaflex are you going to be able to do any more on this?

**Comment by Portaflex** [comment]
I'll try but this is just a modest help. I'm sure many more people would do better.

**Pull request closed without merging**